Add PebbleDB

junha-ahn opened the pull request:
## Proposed changes

- Add PebbleDB

## Types of changes


- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist
- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

## Related issues

...

## Further comments

junha-ahn committed:
add: implement pebbleDB

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/klaytn/klaytn?pullRequest=2145) <br/>All committers have signed the CLA.

blukat29 commented on storage/database/pebbledb_database.go:
```diff
@@ -0,0 +1,318 @@
+}
+
+
+type panicLogger struct{}
+
+func (l panicLogger) Infof(format string, args ...interface{}) {
+}
+func (l panicLogger) Errorf(format string, args ...interface{}) {
+}
+func (l panicLogger) Fatalf(format string, args ...interface{}) {
+	panic(fmt.Errorf("fatal: "+format, args...))
+}
```
There's a module variable `logger` in db_manager.go

```suggestion
type panicLogger struct{}

func (l panicLogger) Infof(format string, args ...interface{}) {
	logger.Info(fmt.Sprintf(format, args...))
}
func (l panicLogger) Errorf(format string, args ...interface{}) {
	logger.Error(fmt.Sprintf(format, args...))
}
func (l panicLogger) Fatalf(format string, args ...interface{}) {
	logger.Crit(fmt.Sprintf(format, args...))
}
```

junha-ahn replied:
Resolved

blukat29 commented:
Please include the pebbleDB to db_manager_test so that we can continuously test while you push commits to this PR branch.

junha-ahn committed:
test: add unit test for PebbleDB

junha-ahn commented:
I added PebbleDB unit test into `db_manager_test.go`. It Success to Run on my local machine. 

But I dont know why CI jobs failed. (with below log) Do I need to do something?

```shell
Error response from daemon: pull access denied for wurstmeister/zookeeper, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
```

junha-ahn commented:
I added PebbleDB unit test into `db_manager_test.go`. and the tests I changed run successfully on my local machine.  However, Im not sure why the CI jobs failed. (with below log) Do I need to do something?

```shell
Error response from daemon: pull access denied for wurstmeister/zookeeper, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
```

> I will also add metrics features soon, and apply lint 

> and If there are any tests that we need to add, plz let me know, I will try to implement them.

hyeonLewis commented:
> I added PebbleDB unit test into `db_manager_test.go`. and the tests I changed run successfully on my local machine. However, Im not sure why the CI jobs failed. (with below log) Do I need to do something?
> 
> ```shell
> Error response from daemon: pull access denied for wurstmeister/zookeeper, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
> ```
> 
There's currently an issue related to zookeeper repo (might be moved). See this 


> I added PebbleDB unit test into `db_manager_test.go`. and the tests I changed run successfully on my local machine. However, Im not sure why the CI jobs failed. (with below log) Do I need to do something?
> 
> ```shell
> Error response from daemon: pull access denied for wurstmeister/zookeeper, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
> ```
> 
> > I will also add metrics features soon, and apply lint
> 
There's an issue related to zookeeper, and will be fixed soon (see https://github.com/klaytn/klaytn/pull/2147)

hyeonLewis commented:
> I added PebbleDB unit test into `db_manager_test.go`. and the tests I changed run successfully on my local machine. However, Im not sure why the CI jobs failed. (with below log) Do I need to do something?
> 
> ```shell
> Error response from daemon: pull access denied for wurstmeister/zookeeper, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
> ```
> 

There's an issue related to zookeeper, and will be fixed soon (see https://github.com/klaytn/klaytn/pull/2147)

junha-ahn committed:
Merge branch 'klaytn:dev' into dev

junha-ahn committed:
add: Metrics

junha-ahn commented:
I added metrics and all the logic from [pebble.go](https://github.com/ethereum/go-ethereum/blob/master/ethdb/pebble/pebble.go#L455) in Geth

junha-ahn commented:
I added metrics and all the logic from [pebble.go](https://github.com/ethereum/go-ethereum/blob/master/ethdb/pebble/pebble.go#L455) in Geth

Do I need to make the metrics key the same as in the [leveldb_database.go](https://github.com/klaytn/klaytn/blob/dev/storage/database/leveldb_database.go#L358) code?

```diff
- d.compWriteMeter = metrics.GetOrRegisterMeter(prefix+"compact/output", nil)
+ db.compWriteMeter = metrics.GetOrRegisterMeter(prefix+"compaction/write", nil) 
```
> "compact/output" from Geth PebbleDB file
> "compaction/write" from Klaytn levelDB file

junha-ahn commented:
NOTE: I have to look up [RocksDB PR](https://github.com/klaytn/klaytn/pull/1855)

junha-ahn commented:
NOTE: I need to refer to the [RocksDB PR](https://github.com/klaytn/klaytn/pull/1855)

hyeonLewis reviewed (commented):
Did you intentionally remove all comments? If not, please bring comments together.

junha-ahn commented:
Ok, I will bring all the comments.

junha-ahn committed:
chore: add comments from Geth

junha-ahn committed:
add: pebbledb cache size option

junha-ahn committed:
Merge remote-tracking branch 'upstream/dev' into dev

junha-ahn committed:
fix: cmd/utils test

junha-ahn committed:
style: apply lint

junha-ahn commented:
Ok, I followed by [LIINT GUIDE](https://github.com/klaytn/klaytn/blob/master/HOW-TO-LINT-YOUR-CHANGE.md) executed below commands:

```shell
klaytn$ gofumpt -w .
klaytn$ goimports -w .
```

When I run `make lint` on my local machine, it prints a lot of lint-bug that I didn't touch. So I dont know how to pass lint-ci 

```shell
$ make lint 

env GOPATH=/Users/junha/go GO111MODULE=on go run build/ci.go lint
>>> /Users/junha/go/bin/golangci-lint run --tests --disable-all --timeout=10m --presets=format --presets=performance ./...
metrics/librato/client.go:94:38: response body must be closed (bodyclose)
        if resp, err = http.DefaultClient.Do(req); err != nil {
                                            ^
metrics/librato/client.go:87:31: should rewrite http.NewRequestWithContext or add (*Request).WithContext (noctx)
        if req, err = http.NewRequest("POST", MetricsPostUrl, bytes.NewBuffer(js)); err != nil {
                                     ^
console/jsre/pretty.go:32: File is not `gci`-ed with --skip-generated -s standard,default (gci)

consensus/istanbul/validator.go:26: File is not `gci`-ed with --skip-generated -s standard,default (gci)
        "github.com/klaytn/klaytn/params"

...
```

> And When I run `cmd/utils/nodecmd/flags_test.go` on my local machine, it fails every time. Do you know anything about it?

hyeonLewis commented:
> I added metrics and all the logic from [pebble.go](https://github.com/ethereum/go-ethereum/blob/master/ethdb/pebble/pebble.go#L455) in Geth
> 
> Do I need to make the metrics key the same as in the [leveldb_database.go](https://github.com/klaytn/klaytn/blob/dev/storage/database/leveldb_database.go#L358) code?

They use same key format for levelDB and pebbleDB. So I think it's also better to use same key format on our side. How do you think @blukat29 ?

blukat29 commented:
@hyeonLewis @junha-ahn Sure, you can use the same metric keys for LevelDB and PebbleDB.

junha-ahn committed:
fix: rename metrics

junha-ahn committed:
Merge remote-tracking branch 'upstream/dev' into dev

blukat29 commented:
Hello @junha-ahn, would you reopen the pull request to https://github.com/kaiachain/kaia repository? we're migrating to kaiachain repos.

junha-ahn commented:
@blukat29 Ok, I will move PR to Kaia by the end of this week. thank you.

junha-ahn commented:
Ok Here is the new PR: https://github.com/kaiachain/kaia/pull/42
